crates/indexer/src/runtime.rs

```diff
@@ -0,0 +1,208 @@
+        "Setting starting version..."
+    );
+    tailer.set_fetcher_version(start_version).await;
+
+    info!(processor_name = processor_name, "Starting fetcher...");
+    tailer.transaction_fetcher.lock().await.start().await;
+
+    let start = chrono::Utc::now().naive_utc();
+
+    info!(
+        processor_name = processor_name,
+        start_version = start_version,
+        "Indexing loop started!"
+    );
+
+    let mut versions_processed: u64 = 0;
+    let mut total_processed: u64 = 0;
+    let mut base: u64 = 0;
+    let mut version_to_check_chain_id: u64 = 0;
+
+    // Check once here to avoid a boolean check every iteration
+    if config.check_chain_id {
+        tailer
+            .check_or_update_chain_id()
+            .await
+            .expect("Failed to get chain ID");
+        version_to_check_chain_id = versions_processed + 100_000;
+    }
+
+    let (tx, mut receiver) = tokio::sync::mpsc::channel(100);
+    let mut tasks = vec![];
+    for _ in 0..config.processor_tasks {
+        let other_tx = tx.clone();
+        let other_tailer = tailer.clone();
+        let task = tokio::task::spawn(async move {
+            loop {
+                let (num_res, res) = other_tailer.process_next_batch().await;
+                other_tx.send((num_res, res)).await.unwrap();
+            }
+        });
+        tasks.push(task);
+    }
+
+    let emit_every = config.emit_every as u64;
+
+    loop {
+        if config.check_chain_id && version_to_check_chain_id < versions_processed {
+            tailer
+                .check_or_update_chain_id()
+                .await
+                .expect("Failed to get chain ID");
+            version_to_check_chain_id = versions_processed + 100_000;
+        }
+
+        let (num_res, result) = receiver
+            .recv()
+            .await
+            .expect("Failed to receive batch results: got None!");
+
+        let processing_result = match result {
+            Ok(res) => res,
+            Err(tpe) => {
+                let (err, start_version, end_version, _) = tpe.inner();
+                error!(
+                    processor_name = processor_name,
+                    start_version = start_version,
+                    end_version = end_version,
+                    error = format!("{:?}", err),
+                    "Error processing batch!"
+                );
+                panic!(
+                    "Error in '{}' while processing batch: {:?}",
+                    processor_name, err
+                );
+            }
+        };
+
+        total_processed += num_res;
+        versions_processed += num_res;
+        if emit_every != 0 {
+            let new_base: u64 = versions_processed / emit_every;
+            if base != new_base {
+                base = new_base;
+                let num_millis =
+                    (chrono::Utc::now().naive_utc() - start).num_milliseconds() as f64 / 1000.0;
+                let tps = (total_processed as f64 / num_millis) as u64;
```

Comment:
Let's change this to tps per batch instead of average lifetime tps.

Reply:
yeah I'll make it a linkedlist